src/jira-client.js

```diff
@@ -1,89 +1,114 @@
+const diagnostics = require('diagnostics');
 const { getCurrentContext } = require('./config');
 
+const debug = {
+  http: diagnostics('jiractl:http'),
+  verbose: diagnostics('jiractl:verbose')
+};
+
 let sessionCookie;
+let authmode = 'basic';
 
 async function getRequestOptions() {
-  if (!sessionCookie) {
+  const opts = {
+    json: true,
+    followAllRedirects: true,
+    headers: {}
+  }
+
+  if (authmode === 'cookie' && !sessionCookie) {
     sessionCookie = await getSessionCookie();
+    opts.headers.Cookie = sessionCookie;
+  } else if (authmode === 'basic') {
+    const { username, password } = getCurrentContext();
+    const encoded = Buffer.from(`${username}:${password}`).toString('base64');
+    opts.headers.Authorization = `Basic ${encoded}`;
   }
 
-  return {
-    json: true,
-    followAllRedirects: true,
-    headers: {
-      Cookie: sessionCookie
-    }
-  };
+  debug.verbose('HTTP Options', opts);
+  return opts;
 }
 
 async function getSessionCookie({
   baseUri = getCurrentContext().uri,
   username = getCurrentContext().username,
   password = getCurrentContext().password } = {}
 ) {
-  try {
-    const { session } = await rp({
-      method: 'POST',
-      uri: makeJiraUri({ baseUri, uri: 'auth/1/session' }),
-      body: {
-        username,
-        password
-      },
-      json: true,
-      followAllRedirects: true
+  const { session } = await rp({
+    method: 'POST',
+    uri: makeJiraUri({ baseUri, uri: 'auth/1/session' }),
+    body: {
+      username,
+      password
+    },
+    json: true,
+    followAllRedirects: true
+  });
+
+  debug.verbose(`New JIRA Session: ${session.name}=${session.value}`)
+  return `${session.name}=${session.value}`;
+}
+
+function makeJiraUri({ baseUri = getCurrentContext().uri, uri, query } = {}) {
+  const fullUri = new URL(`/rest/${ uri }`, `${baseUri}`);
+  if (query) {
+    Object.entries(query).forEach(([key, value]) => {
+      fullUri.searchParams.set(key, value);
     });
-    return `${session.name}=${session.value}`;
-  } catch (err) {
-    throw new Error(err.error.errorMessages ? err.error.errorMessages.join(', ') : err.message);
   }
-}
 
-function makeJiraUri({ baseUri = getCurrentContext().uri, uri } = {}) {
-  return `${baseUri}/rest/${ uri }`;
+  const uriStr = format(fullUri);
+  debug.verbose('Make URI', uriStr);
+  return uriStr;
 }
 
 async function makeQuery({ jql, selector = (results) => results.total } = {}) {
-  const points = getCurrentContext().points;
-  try {
-    const response = await rp(Object.assign({}, await getRequestOptions(), {
-      method: 'POST',
-      uri: makeJiraUri({ uri: 'api/2/search' }),
-      body: {
-        jql,
-        startAt: 0,
-        maxResults: 10000,
-        fields: ['summary', 'status', 'assignee', 'description', points],
-        expand: ['schema', 'names']
-      }
-    }));
-    return selector(response);
-  } catch (err) {
-    throw new Error(err.error.errorMessages.join(', '));
-  }
```

Comment:
The try catch here seems significant.

Reply:
what should I do about that?

Reply:
Wrap your new code in the same try/catch.